### ISSUE TITLE
Docker ENV Private Registry parse error...

### ISSUE THREAD

**Issue by JasonSwindle** [aws/amazon-ecs-agent#122]
Howdy,

Found that if you don't wrap ECS_ENGINE_AUTH_DATA in double quotes on the docker run, you will have a bad time.

### Good
```docker
--env=ECS_ENGINE_AUTH_DATA="{"https://index.docker.io/v1/":{"username":"my_name","password":"my_password","email":"email@example.com"}}"
```

### Bad
```docker
--env=ECS_ENGINE_AUTH_DATA={"https://index.docker.io/v1/":{"username":"my_name","password":"my_password","email":"email@example.com"}}
```

Without double quotes, ECS_ENGINE_AUTH_DATA will be on two lines, and thus not work. Example:

### Bad
```json
"Env": [
    "ECS_LOGFILE=/log/ecs-agent.log",
    "ECS_LOGLEVEL=info",
    "ECS_DATADIR=/data",
    "ECS_CLUSTER=my-cluster",
    "ECS_ENGINE_AUTH_TYPE=dockercfg",
    "ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:auth:REMOVED}",
    "ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:email:REMOVED}",
    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
]
```

### Good
```json
"Env": [
    "ECS_LOGFILE=/log/ecs-agent.log",
    "ECS_LOGLEVEL=info",
    "ECS_DATADIR=/data",
    "ECS_CLUSTER=my-cluster",
    "ECS_ENGINE_AUTH_TYPE=dockercfg",
    "ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:{auth:REMOVED,email:REMOVED}}",
    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
]
```

The error you face when ECS Agent cannot read your ECS_ENGINE_AUTH_DATA looks like:

```bash
2015-06-24T17:22:17Z [ERROR] Unable to decode provided docker credentials module="ecs credentials" type="dockercfg"
```

## TL;DR

The agent should be able to understand the single-line ECS_ENGINE_AUTH_DATA and two line ECS_ENGINE_AUTH_DATA in docker ENV.

Thanks,
Jason

**Comment by euank** [comment]
What you're observing is a bash/zsh expansion really. It's common enough to see something like `ls *.{csv,tsv}`, where the braces expand before the `ls` program ever sees them. What the Docker run command actually sees is the following:

`--env=ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:username:my_name} --env=ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:password:my_password} --env=ECS_ENGINE_AUTH_DATA={https://index.docker.io/v1/:email:email@example.com}`

You notice, bash also destructively removed the double quotes because those also need to be escaped.

If you were to use the `/etc/ecs/ecs.config` file on an ECS Optimized AMI, I do not believe this issue would manifest because bash doesn't enter into the picture there. Similarly, if you used the `--env-file` option and specified these in a file, bash again would not alter these. I also don't think there'd be an issue with the values in a systemd unit file.

Finally, for the 'good' option, it would really be better to use single quotes as they allow bash to escape less. Unless I'm mistaken, the current 'good' one would still strip the double quotes and thus still be invalid JSON.

I see an opportunity for documentation improvement here, but I don't think the agent needs to understand and revert preprocessing done outside of its, and Docker's, control and knowledge.

Does the above make sense, or do you think there is a better option agent-side? I definitely agree this can be a bit confusing and documentation improvements would be helpful.

**Comment by JasonSwindle** [comment]
Howdy,

It 100% does make sense and I agree about the single quotes. This was mainly raised to see if the agent could / would handle it.  Closing. :)
